---
title: "Quality assessment of scRNA seq data"
author: "`r param$author`"
date: "`r format(Sys.time(), '%B, %Y')`"
geometry: "margin=2cm"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    code_folding: "hide"
    highlight: "tango"
    theme: "paper"
bibliography: "`r file.path(param$path_out, 'references.bib')`"
link-citations: yes
csl: "`r file.path(param$path_to_git, 'assets/elsevier-harvard.csl')`"
css: "`r file.path(param$path_to_git, 'assets/style.css')`"
---

```{r r_options, file='../../config/configuration.R'}
#source(file.path(param$path_to_git,'config/configuration.R'))
message("Set R options")
```

```{r rmarkdown, file='../../config/rmarkdown_configuration.R'}
#source(file.path(param$path_to_git,'config/rmarkdown_configuration.R'))
message("Set Rmarkdown options")
```

```{r, include=FALSE, cache=FALSE}
knitr::read_chunk(file.path(param$path_to_git,'scripts/export_data/text.R'))
```

```{r load_packages, warning=FALSE, message=FALSE, results='hide'}

# Load renv and virtualenvs
renv::load(file.path(param$path_to_git,"env/basic"))
renv::use_python(type = "virtualenv", name = file.path(param$path_to_git,"env/basic/virtualenvs/r-reticulate"))
#reticulate::use_virtualenv(file.path(param$path_to_git,"env/basic/virtualenvs/r-reticulate"))

# Required libraries
library(Seurat) # main
library(ggplot2) # plots
library(patchwork) # combination of plots
library(magrittr) # %>% operator
library(knitr)


### Adjustments
# Here only merging!
param$integrate_samples[["method"]]="merge"

```

# Read data

## Read gene annotation
Gene annotation including Ensembl IDs, gene symbols, Entrez Ids, and Seurat gene names, are loaded from a pre-prepared reference file or Ensembl. 

```{r read_gene_annotation, file='../read_data/read_gene_annotation.R'}
#source(file.path(param$path_to_git,'scripts/read_data/read_gene_annotation.R'))
message("Read gene annotation")
```

## Read scRNA-seq data
```{asis, ref.label="read_data"}
```

Here, for the project `r param$project_id`, the following data are analysed:  

```{r read_rds, eval=!is.null(param$data), file='../read_data/read_rds.R', warning=FALSE}
#source(file.path(param$path_to_git,'scripts/read_data/read_rds.R'))
message("Load Seurat object from rds file")
```

```{r restructure_object, eval=!is.null(param$data), warning=FALSE}
# Transfer object into a list
sc_original = sc
sc = list()
n = param$project_id
sc[[n]] = sc_original
```

```{r download_test_dataset, eval=!is.null(param$download_test_datasets), warning=FALSE, message=FALSE}
# Download test dataset 
param$path_test_dataset=paste0(param$path_to_git, "/scripts/download_test_datasets/", param$download_test_datasets, ".R")
if (file.exists(param$path_test_dataset)) {
  message(paste0("Using test dataset '", gsub('download_','', param$download_test_datasets), "'."))
  # Data output in a data subfolder of the directory where it is run 
  # Create output directories
  if (!file.exists("data")) dir.create("data", recursive=TRUE, showWarnings=FALSE)
  setwd(file.path(param$path_to_git,"data"))
  source(param$path_test_dataset)
  param$path_data = data.frame(name=list.dirs(path = file.path(param$path_to_git,"data", "counts"),
                                              full.names = FALSE, recursive = FALSE), 
                                type=c("10x"),
                                path=list.dirs(path = file.path(param$path_to_git,"data", "counts"),
                                              full.names = TRUE, recursive = FALSE))
} else {
  message("Test dataset does not exist.")
}

```

```{r read_datasets, eval=is.null(param$data), warning=FALSE, message=FALSE}
# List of Seurat objects
sc = list()

datasets = param$path_data
for (i in seq(nrow(datasets))) {
  name = datasets[i, "name"]
  type = datasets[i, "type"]
  path = datasets[i, "path"]
  suffix = datasets[i, "suffix"]
  
  # Read 10X or smartseq2
  if (type == "10x") {
    # Read 10X sparse matrix into a Seurat object
    sc = c(sc, ReadSparseMatrix(path, project=name, row_name_column=1, convert_row_names=ensembl_to_seurat_rowname, cellnames_suffix=suffix))
    
  } else if (type == "smartseq2") {
    # Read counts table into a Seurat object
    sc = c(sc, ReadCountsTable(path, project=name, row_name_column=1, convert_row_names=ensembl_to_seurat_rowname, parse_plate_information=TRUE, return_samples_as_datasets=TRUE, cellnames_suffix=suffix))
  } 
}

# Make sure that sample names are unique. If not, just prefix with the dataset name. Also set orig.ident to this name.
sample_names = names(sc)
duplicated_sample_names_idx = which(sample_names %in% sample_names[duplicated(sample_names)])
for (i in duplicated_sample_names_idx) {
  sample_names[i] = paste(head(sc[[i]][["orig.dataset", drop=TRUE]], 1), sample_names[i], sep=".")
  sc[[i]][["orig.ident"]] = sample_names[i]
}

# Set up colors for samples and add them to the sc objects
sample_names = purrr::flatten_chr(purrr::map(sc, function(s) {
  nms = unique(as.character(s[[]][["orig.ident"]]))
  return(nms) 
}))
param$col_samples = GenerateColours(num_colours=length(sample_names), names=sample_names, palette=param$col_palette_samples, alphas=1)
sc = purrr::map(sc, ScAddLists, lists=list(orig.ident=param$col_samples), lists_slot="colour_lists")

message("Read scRNA-seq data into Seurat object")
sc
```

```{r calculate_qc_metadata, file='./calculate_qc_metadata.R', warning=FALSE}
#source(file.path(param$path_to_git,'scripts/pre-processing/calculate_qc_metadata.R'))
message("Calculate and add QC metadata")
```

The following first table shows available metadata (columns) of the first 5 cells (rows). These metadata provide additional information about the cells in the dataset, such as the sample a cell belongs to ("orig.ident"), the number of mapped reads (“nCounts_RNA”), the number of unique genes detected ("`r paste0("nFeature_", param$assay_raw)`"), or percentage of mitochondrial genes ("percent_mt").  
The second table shows available metadata (columns) of the first 5 genes (rows). 
```{r print_metadata}
# Print cell metadata
knitr::kable(head(sc_cell_metadata), align="l", caption="Cell metadata, top 5 rows") %>%
  kableExtra::kable_styling(bootstrap_options=c("striped", "hover"))  %>% 
  kableExtra::scroll_box(width="100%")

# Print gene metadata
knitr::kable(head(sc[[1]][[param$assay_raw]][[]], 5), align="l", caption="Feature metadata, top 5 rows (only first dataset shown)") %>% 
  kableExtra::kable_styling(bootstrap_options=c("striped", "hover"))  %>% 
  kableExtra::scroll_box(width="100%")
```

<br>  
<br>  

# Quality control
Quality control (QC) is an important step of the pre-processing workflow. Here we assess the cell quality, i.e. the cell viability and duplicate rate to determine filter parameter. 

```{asis, ref.label="pre-processing"}
```

## Determining filter thresholds
```{asis, ref.label="qc"}
```

```{r qc_plot_cells, fig.height=fig_patchwork4_height}

# Create plot per QC metric
p_list = list()
for (m in cell_qc_features) {
  p_list[[m]]= ggplot(sc_cell_metadata[, c("orig.ident", m)], aes(x=.data[["orig.ident"]], y=.data[[m]], fill=.data[["orig.ident"]], group=.data[["orig.ident"]])) +
    geom_violin(scale="width")

  # Adds points for samples with less than three cells since geom_violin does not work here
  p_list[[m]] = p_list[[m]] + 
    geom_point(data=sc_cell_metadata[, c("orig.ident", m)] %>% dplyr::filter(orig.ident %in% names(which(table(sc_cell_metadata$orig.ident) < 3))), aes(x=.data[["orig.ident"]], y=.data[[m]], fill=.data[["orig.ident"]]), shape=21, size=2)
  
  # Now add style
  p_list[[m]] = p_list[[m]] + 
    AddStyle(title=m, legend_position="none", fill=param$col_samples, xlab="") + 
    theme(axis.text.x=element_text(angle=45, hjust=1))
  
  # Add filter threshold as segments to plot; min threshold lines are dashed and max threshold lines are twodashed
  if (nrow(cell_qc_thresholds[[m]]) > 0) {
    p_list[[m]] = p_list[[m]] + geom_segment(data=cell_qc_thresholds[[m]], 
                                             aes(x=as.integer(as.factor(orig.ident))-0.5, 
                                                 xend=as.integer(as.factor(orig.ident))+0.5, 
                                                 y=value, yend=value, lty=threshold), colour="firebrick") +
      scale_linetype_manual(values=setNames(c("dashed", "F1"), c("max", "min")))
  }
}
p = patchwork::wrap_plots(p_list, ncol=2) + patchwork::plot_annotation("Distribution of feature values") 
p
```

```{r qc_plot_correlation, fig.height=fig_standard2_height, warning=FALSE}
# Correlate QC metrics for cells
p_list = list()
sc_cell_metadata_plot_order = sample(1:nrow(sc_cell_metadata))

# nFeature vs nCount
m = paste0(c("nCount_", "nFeature_"), param$assay_raw)
p_list[[1]] = ggplot(sc_cell_metadata[sc_cell_metadata_plot_order, , drop=FALSE], aes(x=.data[[m[1]]], y=.data[[m[2]]], colour=.data[["orig.ident"]])) +
  geom_point(size = param$pt_size) + 
  scale_linetype_manual(values=setNames(c("dashed", "F1"), c("max", "min"))) +
  AddStyle(col=param$col_samples)
if (nrow(cell_qc_thresholds[[m[1]]]) > 0) {
  p_list[[1]] = p_list[[1]] + geom_vline(data=cell_qc_thresholds[[m[1]]], aes(xintercept=value, lty=threshold), colour="firebrick")
    
}
if (nrow(cell_qc_thresholds[[m[2]]]) > 0) {
  p_list[[1]] = p_list[[1]] + geom_hline(data=cell_qc_thresholds[[m[2]]], aes(yintercept=value, lty=threshold), colour="firebrick")
}
  

# nFeature vs percent_mt
m = c("percent_mt", paste0(c("nFeature_"), param$assay_raw))
p_list[[2]] = ggplot(sc_cell_metadata[sc_cell_metadata_plot_order, , drop=FALSE], aes(x=.data[[m[1]]], y=.data[[m[2]]], colour=.data[["orig.ident"]])) +
  geom_point(size = param$pt_size) +
  scale_linetype_manual(values=setNames(c("dashed", "F1"), c("max", "min"))) +
  AddStyle(col=param$col_samples)
if (nrow(cell_qc_thresholds[[m[1]]]) > 0) {
  p_list[[2]] = p_list[[2]] + geom_vline(data=cell_qc_thresholds[[m[1]]], aes(xintercept=value, lty=threshold), colour="firebrick")
}
if (nrow(cell_qc_thresholds[[m[2]]]) > 0) {
  p_list[[2]] = p_list[[2]] + geom_hline(data=cell_qc_thresholds[[m[2]]], aes(yintercept=value, lty=threshold), colour="firebrick")
}

# nFeature vs percent_ercc (if available)
if ("percent_ercc" %in% names(cell_qc_features)) {
  m = c("percent_ercc", paste0(c("nFeature_"), param$assay_raw))
  p_list[[3]] = ggplot(sc_cell_metadata[sc_cell_metadata_plot_order, , drop=FALSE], aes(x=.data[[m[1]]], y=.data[[m[2]]], colour=.data[["orig.ident"]])) +
    geom_point(size = param$pt_size) + 
    scale_linetype_manual(values=setNames(c("dashed", "F1"), c("max", "min"))) + 
    AddStyle(col=param$col_samples)
  if (nrow(cell_qc_thresholds[[m[1]]]) > 0) {
    p_list[[3]] = p_list[[3]] + geom_vline(data=cell_qc_thresholds[[m[1]]], aes(xintercept=value, lty=threshold), colour="firebrick")
  }
  if (nrow(cell_qc_thresholds[[m[2]]]) > 0) {
    p_list[[3]] = p_list[[3]] + geom_hline(data=cell_qc_thresholds[[m[2]]], aes(yintercept=value, lty=threshold), colour="firebrick")
  }
}

# Combine plots
p = patchwork::wrap_plots(p_list, ncol=length(p_list)) + patchwork::plot_annotation("QC covariates scatter plots") 
if (length(p_list) == 1) {
  p = p & theme(legend.position="bottom") 
} else {
  p = p + patchwork::plot_layout(guides="collect") & theme(legend.position="bottom") 
}
p
```

<br>  

To save time and computational resources, we __down-sample__ the cell number before continuing with further quality assessment. According to provided number of samples the cell number is down-sampled to 1500 cells (1 sample), 600 (2-3 samples), or 400 (4 and more samples).

```{r downsampling}
# downsample 
# according to samples number to 1000 cells (1 sample), 500 (2-3 samples), or 300 (4 and more samples)

if (length(sc) >= 4) {
  param$downsample_cells_n = 400
} else {
  if (length(sc) > 1) {
    param$downsample_cells_n = 600
  } else {
    param$downsample_cells_n = 1500
  }
}

# downsample_cells_n overwrites downsample_cells_equally
if (!is.null(param$downsample_cells_n)) {
  n = param$downsample_cells_n
} else if (param$downsample_cells_equally) {
  n = purrr::map_int(sc, ncol) %>% min()  
}

# Actual downsampling
if (!is.null(param$downsample_cells_n) | param$downsample_cells_equally) {
  sc = purrr::map(sc, function(s) {
    cells = ScSampleCells(sc=s, n=n, seed=1)
    return(subset(s, cells=cells))
  })
  
  # Adjust combined metadata accordingly
  sc_cell_metadata = sc_cell_metadata[unlist(purrr::map(sc, Cells)), ]
  
  message(paste0("Your data has been down-sampled to ", param$downsample_cells_n, " cells." ))
  print(sc)
}

```

<br>   

## Genes with highest expression
We next plot the genes with the highest median percentage of counts, i.e. for each cell, we calculate the percentage of counts per gene and then calculate the median value of these percentages for each gene in all cells.  
Often those genes are mitochondrial or ribosomal genes. Here rather the percentage of raw counts per gene in a cell and similarity between samples is of importance. A high percentage of of raw counts of those, presumably, less interesting genes, suggest to perform scaling of the data. Differences in percentage of raw counts of those genes between supposedly similar samples might be a indication for a batch effect.  

```{r plot_highest_expression, fig.height=fig_standard_height}
# Plot only samples that we intend to keep 
sc_names = names(sc)[!(names(sc) %in% param$samples_to_drop)]
genes_highestExpr = lapply(sc_names, function(i) {
  top_ten_exp = sc[[i]][[param$assay_raw]][["counts_median"]] %>% dplyr::arrange(dplyr::desc(counts_median)) %>% head(n=10)
  return(rownames(top_ten_exp))
  }) %>%
  unlist() %>%
  unique()

genes_highestExpr_counts = purrr::map_dfc(sc[sc_names], .f=function(s) s[[param$assay_raw]][["counts_median"]][genes_highestExpr, ]) 
genes_highestExpr_counts$gene = genes_highestExpr
genes_highestExpr_counts = genes_highestExpr_counts %>% tidyr::pivot_longer(cols=all_of(sc_names))
genes_highestExpr_counts$name = factor(genes_highestExpr_counts$name, levels=sc_names)

col =  GenerateColours(num_colours=length(genes_highestExpr), names=genes_highestExpr, palette="ggsci::pal_simpsons")
p = ggplot(genes_highestExpr_counts, aes(x=name, y=value, col=gene, group=gene)) + 
  geom_point() + 
  AddStyle(title="Top 10 highest expressed genes per sample, added into one list", 
           xlab="Sample", ylab="Median % of raw counts\n per gene in a cell", 
           legend_position="bottom", 
           col=col)
if (length(unique(genes_highestExpr_counts$name))>1) p = p + geom_line()
p
```

<br>  
<br>  

# Normalization and scaling
```{asis, ref.label="normalization"}
```

```{r normalize, message=FALSE}
# Normalize data the original way
#   This is required to score cell cycle (https://github.com/satijalab/seurat/issues/1679)
if (!("data" %in% sc[[param$project_id]][["RNA"]][])) {
  source(file.path(param$path_to_git,'scripts/pre-processing/normalization.R'))
}

# For testing, perform normalization also for the other method
if (param$norm=="RNA") {
  # Change to perform SCTransform
  param$norm="SCT"
  source(file.path(param$path_to_git,'scripts/pre-processing/normalization.R'))
  # Change back to originally selected standard log normalization
  param$norm="RNA"
  for (i in seq(sc)) {
    DefaultAssay(sc[[i]]) = "RNA"
  }

} else if (param$norm=="SCT") {
  # Change to perform standard log normalization
  for (i in seq(sc)) {
    DefaultAssay(sc[[i]]) = "RNA"
  }
  param$norm="RNA"
  source(file.path(param$path_to_git,'scripts/pre-processing/normalization.R'))
  # Change back to originally selected SCTransform
  param$norm="SCT"
  for (i in seq(sc)) {
    DefaultAssay(sc[[i]]) = "SCT"
  }
}

```

```{r normalization_message}
message(paste0("Here, the ", ifelse(param$norm=="RNA","Standard log normalization","SCTransform")," method was used and no additional sources of variance regressed out."))
```

<br>  

## Variable genes
We select and plot the top 3,000 variable genes. If the composition of the samples is overall similar, it would be expected that they roughly share the same variable genes.
```{asis, ref.label="variable_genes"}
```

```{r plot_variable_features_heights}
fig_height_vf = 5 * ceiling(length(names(sc))/2)
```

```{r plot_variable_features, warning=FALSE, fig.height=fig_height_vf}
# If VariableFeatures not yet present in object
# Find variable features from normalized data (unaffected by scaling)
if (!("scale.data" %in% sc[[param$project_id]][["RNA"]][]) & !param$norm=="SCT") {
  sc = purrr::map(sc, Seurat::FindVariableFeatures, selection.method="vst", nfeatures=3000, verbose=FALSE)
}

# Plot VariableFeaturePlot
p_list = purrr::map(list_names(sc), function(n) {
  top10 = head(Seurat::VariableFeatures(sc[[n]], assay=ifelse(param$norm=="SCT", param$norm, param$assay_raw)), 10)
  p = Seurat::VariableFeaturePlot(sc[[n]], 
                                  assay=ifelse(param$norm=="SCT", param$norm, param$assay_raw), 
                                  selection.method=ifelse(param$norm=="RNA", "vst", "sct"), 
                                  col=c("grey", param$col), pt.size = param$pt_size) + 
    AddStyle(title=n) + 
    theme(legend.position=c(0.2, 0.8), legend.background=element_rect(fill=alpha("white", 0.0)))
  p = LabelPoints(plot=p, points=top10, repel=TRUE, xnudge=0, ynudge=0)
  return(p)
})

p = patchwork::wrap_plots(p_list, ncol=2) + patchwork::plot_annotation("Variable genes")
p
```

If multiple data sets are processed, we merge the datasets now.  
```{r combine_datasets}
if (length(sc) == 1) {
  # Default assay is set automatically
  sc = sc[[1]]
  message("Your dataset contains 1 sample only. No merging/integrating required.")
} else {
  source(file.path(param$path_to_git,'scripts/pre-processing/combine_samples.R'))
  message("In this workflow, sample combination is always performed via dataset merging as merged object is required for streamlined integration.")
}


# Only relevant if data loaded from rds, otherwise they would be scaled and dimensional reduced before
if (!("scale.data" %in% sc[["RNA"]][])) {
  # Scale (default)
  all_genes = rownames(sc)
  sc = suppressMessages(ScaleData(sc, features = all_genes))
} 
# Run pca with the variable genes
if (!("pca" %in% list_names(sc@reductions[]))) {
  # Run PCA for default normalization
  sc = suppressMessages(Seurat::RunPCA(sc, features=Seurat::VariableFeatures(object=sc), verbose=FALSE, npcs=min(50, ncol(sc))))
}

```

<br>  

## Relative log expression
To better understand the efficiency of the applied normalization procedures, we plot the relative log expression of genes in 100 randomly selected cells per sample before and after normalization. This type of plot reveals unwanted variation in your data.  
The concept is taken from `r knitcitations::citet("10.1371/journal.pone.0191629")`. In brief, we remove variation between genes by calculating for each gene its median expression across all cells, and then calculate the deviation from this median for each cell. For each cell, we plot the median expression (black), the interquartile range (<span style="color:lightgrey;font-weight:bold">lightgrey</span>), whiskers defined as 1.5 times the interquartile range (<span style="color:darkgrey;font-weight:bold">darkgrey</span>), and outliers (`r paste0('<span style="color:', param$col_samples, ';font-weight:bold">', param$col_samples, '</span>', collapse=', ')`).  
If expression levels of most genes are supposed to be similar in all cell types, sample heterogeneity is a sign of unwanted variation.  
```{r plot_RLE_raw, fig.height=fig_standard2_height}

n_cells_rle_plot = 100

# Sample at most 100 cells per dataset and save their identity
cells_subset = sc[["orig.ident"]] %>% tibble::rownames_to_column() %>% 
  dplyr::group_by(orig.ident) %>% 
  dplyr::sample_n(size=min(n_cells_rle_plot, length(orig.ident))) %>% 
  dplyr::select(rowname, orig.ident)

# Plot raw data
p1 = PlotRLE(as.matrix(log2(GetAssayData(subset(sc, cells=cells_subset$rowname), assay=param$assay_raw, layer = "counts") + 1)), 
            id=cells_subset$orig.ident, 
            col=param$col_samples) + 
  labs(title="log2(raw counts + 1)")

p2 = PlotRLE(as.matrix(GetAssayData(subset(sc, cells=cells_subset$rowname), assay="RNA", layer = "data")), 
            id=cells_subset$orig.ident, 
            col=param$col_samples) + 
  labs(title="Standard logNormalised data")

p3 = PlotRLE(as.matrix(GetAssayData(subset(sc, cells=cells_subset$rowname), assay="SCT", layer = "data")), 
            id=cells_subset$orig.ident, 
            col=param$col_samples) + 
  labs(title="SCTransformed data")

p = p1 / p2 / p3
p
```

<br>  
<br>  

# Investigating covariants

```{asis, ref.label="regress_out"}
```

<br>  

## Dimensional reduction
```{asis, ref.label="dimensional_reduction"}
```

<br>  

## Principal component analysis
Running a PCA on our object, we see how the variance can be explained.  
In case a early PC dimension is split on cell-cycle genes or mitochondrial gene, we might want to regress this signal from the data, so that cell-cycle heterogeneity or expression of mitochondrial genes does not contribute to PCA or downstream analysis. However, remember, those signals could also always be informative of the biology!  
```{r pca_loadings, fig.height=14}
p_list = Seurat::VizDimLoadings(sc, dims=1:12, reduction="pca", col=param$col, combine=FALSE, balanced=TRUE)
for (i in seq(p_list)) p_list[[i]] = p_list[[i]] + AddStyle(xlab = paste0("PC ",i))
p =  patchwork::wrap_plots(p_list, ncol=4) + patchwork::plot_annotation("Top gene loadings of the first two PCs") 
p
```

<br>  

## Feature plots
The feature plots allow to examined the distribution of cells with specific features, here QC covariates (Number of counts and features as well as percent of mitochondrial and ribosomal reads), and, hence, to infer whether those feature might exert some influence on the localisation and distribution of cells in low dimensional space. These plot can clarify the necessity and scope of covariates filter application.  
```{r umap, message=FALSE}
# Generate UMAP
if (!("umap" %in% list_names(sc@reductions[]))) {
  # Default UMAP
  sc = suppressWarnings(Seurat::RunUMAP(sc, dims=1:param$pc_n, verbose=FALSE, umap.method="uwot", n.neighbors=param$umap_k))
}

```

```{r feature_plots, fig.height=fig_standard4_height}

# Number of counts 
qc_feature = paste0("nCount_", param$assay_raw)
p1 = suppressMessages(Seurat::FeaturePlot(sc, features=qc_feature, pt.size = param$pt_size) + 
  AddStyle(title="Number of counts", legend_position="right", xlab = "UMAP 1", ylab = "UMAP 2") + 
  scale_colour_gradient(low=param$col_bg, high=param$col, trans="log10"))

# Number of features
qc_feature = paste0("nFeature_", param$assay_raw)
p2 = suppressMessages(Seurat::FeaturePlot(sc, features=qc_feature, pt.size = param$pt_size) + 
  AddStyle(title="Number of features", legend_position="right", xlab = "", ylab = "") + 
  scale_colour_gradient(low=param$col_bg, high=param$col, trans="log10"))

# Percent mitochondrial reads
p3 = Seurat::FeaturePlot(sc, features="percent_mt", cols=c(param$col_bg, param$col), pt.size = param$pt_size) + 
  AddStyle(title="% mitochondrial", legend_position="right", xlab = "", ylab = "")

# Percent ribosomal reads
p4 = Seurat::FeaturePlot(sc, features="percent_ribo", cols=c(param$col_bg, param$col), pt.size = param$pt_size) +
  AddStyle(title="% ribosomal", legend_position="right", xlab = "", ylab = "")

#p = ((p2 / p3) | (p4 / p5)) + 
#    plot_layout(widths = c(2, 2))
p = p1 | p2 | p3 | p4
p

```

<br>  

## Cell Cycle Effect
The following plots can help to evaluate whether the cell cycle phases are a major source of variation in the dataset and the necessity to regress out cell cycle effects during scaling.  
An indication to consider removal of cell cycle effects would be a strong differences between samples (that are supposed to have a similar cell composition) or if cells in the same cell cycle phase form very distinctive clusters. Nevertheless, the final decision also depends on the biological system and scientific question.  

```{asis, ref.label="cc-removal"}
```

```{r cellCycleEffect, fig.height=fig_standard2_height}

# Set up colours for cell cycle effect and add to sc object
col =  GenerateColours(num_colours=length(levels(sc$Phase)), names=levels(sc$Phase), palette="ggsci::pal_npg", alphas=1)
sc = ScAddLists(sc, lists=list(Phase=col), lists_slot="colour_lists")

# Plot umap colored by cell cycle phase
p1 = Seurat::DimPlot(sc, reduction="umap", group.by="Phase", pt.size = param$pt_size, cols = col) +
  AddStyle(title="", xlab = "UMAP 1", ylab = "UMAP 2") +
  NoLegend()

# Plot umap colored by cell cycle phase
p2 = Seurat::DimPlot(sc, reduction="umap", group.by="Phase", split.by = "Phase", pt.size = param$pt_size, cols = col, ncol = 1) +
  AddStyle(title="", xlab = "UMAP 1", ylab = "UMAP 2") +
  NoLegend()

# Fraction of cells per cell cycle phase
p3 = ggplot(sc[[]] %>% 
              dplyr::group_by(orig.ident, Phase) %>% 
              dplyr::summarise(num_cells=length(Phase)), 
            aes(x=orig.ident, y=num_cells, fill=Phase)) + 
  geom_bar(stat="identity", position="fill") + 
  scale_y_continuous("Fraction of cells") +
  AddStyle(fill=Misc(sc, "colour_lists")[["Phase"]]) + 
  theme(axis.text.x = element_text(angle=30, hjust=1, vjust=1, size = 12)) + xlab("")

p = p1 + p2 + plot_spacer() + p3 + 
  plot_layout(widths = c(3, 1, 1, 1)) + 
  patchwork::plot_annotation(title="Cell cycle phases") 
p
```

<br>  

## Expression of apoptosis related genes 
Here the expression of some apoptosis related genes is plotted to get, apart from the mitochondrial gene expression ratio, another indication of cell viability. However, the expression level of such genes strongly depends on the state during the process of apoptosis, i.e. low expression is not necessarily affirmation of high cell viability. 
```{r apoptosis, fig.height=fig_patchwork8_height, warning=FALSE}

# Feature plots of target genes
if (param$species=="human") {
  apo_genes = c("BCL2", "BIRC5", "CASP3", "TP53")
} else {
  apo_genes = c("Bcl2", "Birc5", "Casp3", "Tp53")
}

p1 = suppressMessages(Seurat::FeaturePlot(sc, features=apo_genes, cols=c(param$col_bg, param$col), pt.size = param$pt_size, combine = FALSE))
p1 = AddStyleMultiFeaturePlot(p1, title = 'UMAP projection of gene expression', ncol = 4)

p2 = suppressMessages(Seurat::VlnPlot(sc, features=apo_genes, group.by = "orig.ident", slot = "counts", log = TRUE, cols = param$col_samples, combine = FALSE))
p2 = AddStyleMultiVln(p2, title = 'Expression levels', ncol = 4)

p = p1 / p2 
p
```

<br>  
<br>  


# Batch correction 
Integration can help to remove biological differences and match shared cell types and states across datasets.   
```{asis, ref.label="sample_combination"}
```

<br>  

### Methods for sample combination {.tabset}
```{asis, ref.label="integration_methods"}
```
  
<br>  

To explore the results of data merging and different integration methods, here integration is performed in low-dimensional space (streamlined (one-line) integrative analysis as described in https://satijalab.org/seurat/articles/seurat5_integration).

<details>
  <summary>How does streamlined (one-line) integrative analysis work?</summary>
  
  Seurat v5 enables streamlined integrative analysis via the IntegrateLayers function.   
  First, the samples are merged and then the layers are split (a counts and data layer for each batch). As the the data is split into layers, normalization and variable feature identification is performed for each set independently and a consensus set of variable features is identified. Moreover, a dimensional reduction with (hopefully) co-embed shared cell types across samples is returned as output.  
</details>

The following plots offer a low dimension representation of your data combined via the different methods. 

```{r streamlined_integrative_analysis, file='./streamlined_integrative_analysis.R', warning=FALSE, message=FALSE}
#source(file.path(param$path_to_git,'scripts/pre-processing/streamlined_integrative_analysis.R'))
```

```{r dim_reductions}
### Score plots
# PCA score plots colored by sample
p1 = Seurat::DimPlot(sc, reduction="pca", group.by = "orig.ident", cols=param$col_samples, pt.size = param$pt_size, dims = c(1,2)) + 
  AddStyle(legend_position="none", title = "", xlab = "PC 1", ylab = "PC 2")
p2 = Seurat::DimPlot(sc, reduction="pca", group.by = "orig.ident", cols=param$col_samples, pt.size = param$pt_size, dims = c(3,4)) + 
  AddStyle(legend_position="none", title = "", xlab = "PC 3", ylab = "PC 4")

# CCA integrated score plots colored by sample
p3 = Seurat::DimPlot(sc, reduction="integrated.cca", group.by = "orig.ident", cols=param$col_samples, pt.size=param$pt_size, dims = c(1,2)) + 
  AddStyle(legend_position="none", title = "", xlab = "PC 1", ylab = "PC 2")
p4 = Seurat::DimPlot(sc, reduction="integrated.cca", group.by = "orig.ident", cols=param$col_samples, pt.size=param$pt_size, dims = c(3,4)) + 
  AddStyle(legend_position="none", title = "", xlab = "PC 3", ylab = "PC 4")

# RPCA integrated score plots colored by sample
p5 = Seurat::DimPlot(sc, reduction="integrated.rpca", group.by = "orig.ident", cols=param$col_samples, pt.size=param$pt_size, dims = c(1,2)) + 
  AddStyle(legend_position="none", title = "", xlab = "PC 1", ylab = "PC 2")
p6 = Seurat::DimPlot(sc, reduction="integrated.rpca", group.by = "orig.ident", cols=param$col_samples, pt.size=param$pt_size, dims = c(3,4)) + 
  AddStyle(legend_position="none", title = "", xlab = "PC 3", ylab = "PC 4")

### Umaps
# Plot umap colored by sample
p7 = Seurat::DimPlot(sc, reduction="umap", group.by="orig.ident", pt.size = param$pt_size, cols = param$col_samples) +
  AddStyle(title="", legend_position="right", xlab = "UMAP 1", ylab = "UMAP 2")

# Plot CCA umap colored by sample
p8 = Seurat::DimPlot(sc, reduction="umap.cca", group.by="orig.ident", pt.size = param$pt_size, cols = param$col_samples) +
  AddStyle(title="", legend_position="right", xlab = "UMAP 1", ylab = "UMAP 2")

# Plot RPCA umap colored by sample
p9 = Seurat::DimPlot(sc, reduction="umap.rpca", group.by="orig.ident", pt.size = param$pt_size, cols = param$col_samples) +
  AddStyle(title="", legend_position="right", xlab = "UMAP 1", ylab = "UMAP 2")

# Combine plots
p_list = list()
p_list[[1]] = (p1 / p2)
p_list[[1]] = (p_list[[1]] | p7) +
  plot_layout(widths = c(1, 3)) + 
  plot_annotation(title = 'Merged') 
p_list[[2]] = (p3 / p4)
p_list[[2]] = (p_list[[2]] | p8) +
  plot_layout(widths = c(1, 3)) + 
  plot_annotation(title = 'CCA integrated') 
p_list[[3]] = (p5 / p6)
p_list[[3]] = (p_list[[3]] | p9) +
  plot_layout(widths = c(1, 3)) + 
  plot_annotation(title = 'RPCA integrated') 

```

#### Merged
```{r dim_reductions_merged, fig.height=fig_patchwork8_height}
p_list[[1]]
```

<br>   
<br>  

#### CCA integrated
```{r dim_reductions_cca_integrated, fig.height=fig_patchwork8_height}
p_list[[2]]
```

<br>   
<br>  

#### RPCA integrated
```{r dim_reductions_rpca_integrated, fig.height=fig_patchwork8_height}
p_list[[3]]
```

<br>   
<br>  


# Determinig dimensionality of the dataset
Now we need to decide how many PCs we want to use for our analyses.  

```{asis, ref.label="determinig_dimensionality"}
```

```{r dimensionality, fig.height=fig_standard_height}
# More approximate technique used to reduce computation time
p = Seurat::ElbowPlot(sc, ndims=min(20, ncol(sc))) + 
  geom_vline(xintercept=param$pc_n + .5, col="firebrick", lty=2) + 
  AddStyle(title="Elbow plot") 
p

# Cannot have more PCs than number of cells
param$pc_n = min(param$pc_n, ncol(sc))
```


<br>
<br>
<br>

```{r, child='../export_data/appendix.Rmd'}
```

