
<br>

# Data export

```{r postprocess_sc, warning=FALSE, results="hide"}
### Add metadata to misc
# Add colour lists for orig.dataset
col = GenerateColours(num_colours=length(levels(sc$orig.dataset)), names=levels(sc$orig.dataset), palette=param$col_palette_samples, alphas=1)
sc = ScAddLists(sc, lists=list(orig.dataset=col), lists_slot="colour_lists")

# Add experiment details
Seurat::Misc(sc, "experiment") = list(project_id=param$project_id, date=Sys.Date(), species=gsub("_gene_ensembl", "", param$mart_dataset))

# Add parameter
Seurat::Misc(sc, "parameters") = param

# Add technical output (note: cannot use Misc function here)
sc@misc$technical = data.frame(ScrnaseqSessionInfo(param$path_to_git))
```

<details>
  <summary> Export seurat object</summary>
  
  We export the assay data, cell metadata, clustering and visualization.   
    
  Result files:  
  * sc.rds: Seurat object for import into R  
  * cell_metadata.xlsx: Cell metadata in excel format  
  * counts folder: Contains raw count matrix of the entire dataset (sparse matrix)  

```{r output_files, message=FALSE, warning=FALSE, results="hide"}
### Export seurat object
saveRDS(sc, file=file.path(param$path_out, "data", "sc.rds"))

### Export count matrix
invisible(ExportSeuratAssayData(sc, 
                                dir=file.path(param$path_out, "data", "counts"), 
                                assays=param$assay_raw, 
                                slot="counts",
                                include_cell_metadata_cols=colnames(sc[[]]),
                                metadata_prefix=paste0(param$project_id, ".")))

### Export metadata
openxlsx::write.xlsx(x=sc[[]] %>% tibble::rownames_to_column(var="Barcode"), file=file.path(param$path_out, "data", "cell_metadata.xlsx"), rowNames=FALSE, colNames=TRUE)

```

</details>


<details>
  <summary>Export as andata object</summary>
  
  We export the assay data, cell metadata, clustering and visualization in andata format.  
    
  Result file:  
  * sc.h5ad: H5AD object   

```{r andata_export, message=FALSE, warning=FALSE, results="hide"}
# Convert Seurat v5 single cell object to anndata object
scCustomize::as.anndata(x = sc, file_path = file.path(param$path_out, "data"), file_name = "sc_anndata.h5ad", )

```

</details>


<details>
  <summary> Export to Loupe Cell Browser</summary>
  
  If all provided datasets are of type "10x", we export the UMAP 2D visualization, metadata such as the cell clusters, and lists of differentially expressed genes, so you can open and work with these in the Loupe Cell Browser (https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser).  
    
  Result files are:  
  * Loupe_projection_(umap|pca|...).csv: Seurat UMAP/PCA/... projections for visualization  
  * Loupe_metadata.csv: Seurat cell meta data including clusters and cell cycle phases  
  * Loupe_genesets.csv: Gene sets such as markers, DEGs, cell cycles   
    
  Projections can be imported in Loupe via "Import Projection", cell meta data via "Import Categories" and gene sets via "Import Lists".  

```{r loupe_export, warning=FALSE, results="hide"}
### Export to Loupe Cell Browser
if (all(param$path_data$type == "10x")) { 
  
  # Export reductions (umap, pca, others)
  for(r in Seurat::Reductions(sc)) {
    write.table(Seurat::Embeddings(sc, reduction=r)[,1:2] %>% as.data.frame() %>% tibble::rownames_to_column(var="Barcode"),
                file=file.path(param$path_out, "data", paste0("Loupe_projection_", r, ".csv")), col.names=TRUE, row.names=FALSE, quote=FALSE, sep=",")
  }
  
  # Export categorical metadata
  loupe_meta = sc@meta.data
  idx_keep = sapply(1:ncol(loupe_meta), function(x) !is.numeric(loupe_meta[,x]))
  write.table(x=loupe_meta[, idx_keep] %>% tibble::rownames_to_column(var="barcode"), 
              file=file.path(param$path_out, "data", "Loupe_metadata.csv"), col.names=TRUE, row.names=FALSE, quote=FALSE, sep=",")
  
  # Export gene sets (CC genes, known markers, per-cluster markers up- and downregulated, ...)
  gene_lists = Misc(sc, "gene_lists")
  
  # Remove empty gene sets
  gene_lists = gene_lists[purrr::map_int(gene_lists, length) > 0]
  loupe_genesets = purrr::map_dfr(names(gene_lists), function(n) {return(data.frame(List=n, Name=gene_lists[[n]]))})
  loupe_genesets$Ensembl = seurat_rowname_to_ensembl[loupe_genesets$Name]
  write.table(loupe_genesets, file=file.path(param$path_out, "data", "Loupe_genesets.csv"), col.names=TRUE, row.names=FALSE, quote=FALSE, sep=",")
}

```

</details>



